Sources/NIOSOCKS/Messages/ClientRequest.swift

```diff
@@ -0,0 +1,187 @@
+    
+    case domain(String, port: UInt16)
+    
+    /// How many bytes are needed to represent the address, excluding the port
+    var size: Int {
+        switch self {
+        case .address(.v4):
+            return 4
+        case .address(.v6):
+            return 16
+        case .address(.unixDomainSocket):
+            fatalError("Unsupported")
+        case .domain(let domain, port: _):
+            return domain.count + 1
+        }
+    }
+}
+
+extension ByteBuffer {
+    
+    mutating func readAddresType() throws -> AddressType? {
+        let save = self
+        guard let type = self.readInteger(as: UInt8.self) else {
+            self = save
+            return nil
+        }
+        
+        switch type {
+        case 0x01:
+            return try self.readIPv4Address()
+        case 0x03:
+            return try self.readDomain()
+        case 0x04:
+            return try self.readIPv6Address()
+        default:
+            throw SOCKSError.InvalidAddressType(actual: type)
+        }
+    }
+    
+    mutating func readIPv4Address() throws -> AddressType? {
+        let save = self
+        guard
+            let bytes = self.readSlice(length: 4),
+            let port = try self.readPort()
+        else {
+            self = save
+            return nil
+        }
+        return .address(try .init(packedIPAddress: bytes, port: port))
+    }
+    
+    mutating func readIPv6Address() throws -> AddressType? {
+        let save = self
+        guard
+            let bytes = self.readSlice(length: 16),
+            let port = try self.readPort()
+        else {
+            self = save
+            return nil
+        }
+        return .address(try .init(packedIPAddress: bytes, port: port))
+    }
+    
+    mutating func readDomain() throws -> AddressType? {
+        let save = self
+        guard
+            let length = self.readInteger(as: UInt8.self),
+            let host = self.readString(length: Int(length)),
+            let port = try self.readPort()
+        else {
+            self = save
+            return nil
+        }
+        return .domain(host, port: UInt16(port))
+    }
+    
+    mutating func readPort() throws -> Int? {
+        return self.readInteger(as: UInt16.self).flatMap { Int($0) }
+    }
+    
+    @discardableResult mutating func writeAddressType(_ type: AddressType) -> Int {
+        switch type {
+        case .address(.v4(let address)):
+            return self.writeInteger(UInt8(1))
+                + self.writeInteger(address.address.sin_addr.s_addr, endianness: .little)
+                + self.writeInteger(address.address.sin_port, endianness: .little)
+        case .address(.v6(let address)):
+            return self.writeInteger(UInt8(4))
+                + self.writeIPv6Address(address.address)
+                + self.writeInteger(address.address.sin6_port, endianness: .little)
+        case .address(.unixDomainSocket):
+            fatalError("unsupported")
```

Comment:
Can we comment here that we don't expect to hit this branch because we validate elsewhere?